src/swfbytes.1
.TH swfbytes "1" "April 2012" "swfbytes" "swftools"
.SH NAME
swfbytes \- Tool for simplifying the hex-editing of SWF files.

.SH Synopsis
.B swfbytes [\-v] file.swf > file.hexdump
.PP
.B swfbytes file.hexdump

.SH DESCRIPTION
With this tool and a text editor, you can perform some basic
editing operations on SWF files.

.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR 
    Print help and exit
.TP
\fB\-v\fR, \fB\-\-verbose\fR 
    Be more verbose
.TP
\fB\-V\fR, \fB\-\-version\fR 
    Print program version and exit
.SH AUTHOR

Matthias Kramm <kramm@quiss.org>


ChangeLog
0.0.1: initial release
0.0.2: 
	* pdf2swf: Added support for pdf bitmaps, clipping, state saving/restoring, colors,
	           solid fill and line width. The output SWF files are now much smaller.
	* swfcombine: Added stacking option (-t)
	* swfcombine: Small bugfixes with the command line options
	* rfxswf-lib: some bugfixes concerning sprites, cgi, shapes, jpegs.
	* swfstrings: New: Allows extraction of text from swf files
	* swfdump: New: Dumps swf file information. (And does some error checking on the SWF)
	* jpeg2swf: New: Converts one or more jpegs to a swf movie.

0.1.0: 
	* configure: Now works without automake, automatically disables 
			jpeg2swf and pdf2swf if jpeglib is not found (disable
			both) or t1lib is not found (disable pdf2swf)
	* swfcombine: bugfix: DEFINETEXT tags didn't get relocated right
	* swfcombine: bugfix: no longer builds "unstreamable" SWFs
	* rfxswf-lib: actionscript
	* rfxswf-lib: lossless image formats
	* pdf2swf: conversion of lossless image formats
	* pdf2swf: picture cache to prevent images getting stored more than once
	* pdf2swf: conversion of (hyper)links
	* swfdump: -a, -e and -XY options
	* lib/example/zlibtest: new file, demonstrates using lossless images with rfxswf-lib
	* lib/example/dumpfont: utility for dumping fonts out of swf (and generate c sourcecode for creating them)
	* lib/example/demofont: example font, generated with lib/example/dumpfont.

0.1.1:  
	* swfcombine: Wasn't able to handle converted pdfs with links. Fixed.
	* swfcombine: Introduce --dummy, -X,-Y,-r options
	* swfdump: new: -r option
	* pdf2swf: small bugfixes
	* HOWTO_pdf2swf: added section about correcting size and framerate

0.1.2:
	* swfdump: can dump all actionscript flavours (button, button2, standalone) now
	* rfxswf-lib: improved font handling
	* lib/example/makefonts.cc: new file, converts the standard fonts to swf
	* lib/example/glyphshape.c: new file, demonstrates using font characters as shapes
	* pdfswf: many bugfixes. The "pdfscreen" manual now get's converted correctly
	* rfxswf-lib: bugfix: Couldn't handle jpeg grayscale images

0.2.0:	
	* pdf2swf: TrueType support added. (using ttf2pt1)
	* rfxswf-lib: performs conversion of jpeg colorspaces now. (Makes pdf2swf cope
	               with more types of jpegs)
	* swfextract: new tool, allows to extract single objects, frames etc. from movieclips
	* swfcombine: allows ids instead of names, now, too.
	* swfdump: added text displaying functionality. (like swfstrings)
	* swfdump: now dumps also the undocumented placeobject2 actioncode.
	* lib/example/avi2swf.cc: For now, converts small (!) avi movie files to swf.  (using 
			          libavifile)

0.2.1:	
	* configure: Does checking for byte order and command substitution capability now.
	* rfxswf: Applied Anjo Krank's patch for MacOS X.
	* pdf2swf: Fixed font transformation matrix
	* swfcombine: new options: --cat, --stack1, --merge, --frame
	* swfs/: new directory, holds some standard SWFs primarily used for pdf2swf
	         installs into ${DATADIR}/swftools/swfs
	* pdf2swf: New options --viewer, --preloader, -b, -l
	* pdf2swf: Fonts are now in ${DATADIR}/swftools/fonts

0.2.2:	
	* swfs/: Added/Wrote sources for some of the swf files 
	* swfdump: Major bugfix in placeobject2, hex output function
	* swfextract: Can now extract png, jpeg and mp3.
	* swfcombine: big endian fixes
	* swfextract: big endian fixes
	* pdf2swf: Changed from 100dpi to 72dpi conversion
	* pdf2swf: Some optimizations, fix the zero string problem.
	* lib/: Added Bladeenc 0.94.1 for mp3 compression
	* lib/example/avi2swf.cc: Can now handle big files. Slightly better compression ratio. Sound.
	* lib/example/sound.c: New file, demonstrates use of the sound module in librfxswf.

0.2.3: 
	* pdf2swf: New option --fontdir, can now handle external fonts.
	* lib/modules/swftext.c: Several font-extraction bugfixes
	* src/swfdump: Bugfixes, ids and depths are now in decimal
	* src/swfcombine: Bugfixes.
	* configure.in: Cleaned up.

0.3.0:
	* pdf2swf: The --ignore option now actually makes the pdf files smaller.
	* rfxswf: Can now write DefineFont2 tags
	* swfextract/swfcombine: Bugfixes (DefineEditText)
	* swfdump: New option --used, prints out referred IDs of each tag.
	* rfxswf: Can now read and write Flash 6 (MX) compressed SWFs. (SWCs)
	* pdf2swf: Added compression flag (-z)
	* swfcombine: Added compression flag (-z)
	* configure.in: zlib is now required.
	* src/png2swf: New tool.
	* swfdump: Prints some image infos now.
	* pdf2swf: Fixed Image handling
	* avi2swf/: Is part of the main distribution now
	* src/wav2swf: New tool.

0.3.1:
	* configure: Didn't create all Makefiles when t1lib or jpeglib were absent. Fixed.
	* swfs/Makefile: FreeBSD fixes.

0.4.0:
	* pdf2swf/xpdf: Upgraded PDF Parser to Version 1.01.
	* pdf2swf: Can now handle Type 3 Fonts. (As Bitmaps)
	* src/swfcombine: Completely rewritten, using the rfxswf routines
	* configure: Check for libtool
	* lib/rfxswf: Fixed some parsing bugs (swffont, swfdump)
	* pdf2swf: Added antialiasing capabilities for Type 3 Fonts
	* lib/examples/makefonts: Fonts: Bounding boxes are now handled correctly
	* png2swf: Support for more image types
	* pdf2swf: New option -t (--stop)

0.4.1:
	* src/swfcombine: Bugfix with -x option
	* lib/Makefile: MacOSX fixes
	* pdf2swf/SWFOutput.cc: fixed undefined sqrt()
	* src/png2swf: Support for transparency
	* pdf2swf: Support for CID Fonts
	* src/swfcombine: Bugfixes
	* lib/rfxswf: Added Sprite Unfolding routines

0.4.2:
	* src/Makefile: OpenBSD fixes
	* lib/lame: Removed *.o files from archive
	* avi2swf: Implemented --flip option
	* pdf2swf: Implemented --flashversion option
	* pdf2swf: Implemented Viewer gadgets- it's now possible to call Flash functions via hyperlink

0.4.3:
	* src/jpeg2swf: Grayscale fix
	* lib/rfxswf: Improved MX support
	* pdf2swf: Command line fixes
	* swfdump: Added -f (--frames) option
	* png2swf: Added support for mode 6 images
	* swfs: Support for print protection in simple_viewer

src/swfbytes.c
/* swfbytes.c
   A tool for modifying swfs on the tag level

   Part of the swftools package.
   
   Copyright (c) 2008/2009 Matthias Kramm <kramm@quiss.org>
 
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */

#include "../config.h"
#include <unistd.h>
#include <stdio.h>
#include <fcntl.h>
#include <stdarg.h>
#include "../lib/rfxswf.h"
#include "../lib/args.h"

static char * filename = 0;
static int verbose;
static char* output_filename = "output.swf";

static struct options_t options[] = {
{"h", "help"},
{"v", "verbose"},
{"V", "version"},
{0,0}
};

int args_callback_option(char*name,char*val)
{
    if(!strcmp(name, "V")) {
        printf("swfedit - part of %s %s\n", PACKAGE, VERSION);
	return 0;
    } else if(!strcmp(name, "v")) {
	verbose++;
	return 0;
    } else if(!strcmp(name, "o")) {
	output_filename = val;
	return 1;
    } else {
        printf("Unknown option: -%s\n", name);
	return 0;
    }
    return 0;
}
int args_callback_longoption(char*name,char*val)
{
    return args_long2shortoption(options, name, val);
}
void args_callback_usage(char *name)
{
    printf("\n");
    printf("Usage: %s [-v] file.swf > file.hexdump\n", name);
    printf("OR:    %s file.hexdump\n", name);
    printf("\n");
    printf("-h , --help                    Print help and exit\n");
    printf("-v , --verbose                 Be more verbose\n");
    printf("-V , --version                 Print program version and exit\n");
    printf("\n");
}
int args_callback_command(char*name,char*val)
{
    if(filename) {
        fprintf(stderr, "Only one file allowed. You supplied at least two. (%s and %s)\n", filename, name);
    }
    filename = name;
    return 0;
}

void dumpTag(FILE*fo, char*prefix, TAG*tag)
{
    int t;
    for(t=0;t<tag->len;t++) {
	if(!(t&15))
	    fprintf(fo, "%s| ", prefix);
	fprintf(fo, "%02x ", tag->data[t]);
	if((t && ((t&15)==15)) || (t==tag->len-1)) {
	    fprintf(fo, "\n");
	}
    }
}

void dumpFile(SWF*swf, FILE*fo)
{
    TAG* tag = swf->firstTag;
    int indent = 0;
    int t;
    char whitespace[33];
    char*prefix = "";

    memset(whitespace, 32, 32); whitespace[32] = 0;
    
    fprintf(fo, "Version: %d\n", swf->fileVersion);
    fprintf(fo, "FrameRate: %f\n",swf->frameRate/256.0);
    fprintf(fo, "FrameCount: %d\n",swf->frameCount);
    fprintf(fo, "Width: %.2f\n",(swf->movieSize.xmax-swf->movieSize.xmin)/20.0);
    fprintf(fo, "X-Offset: %.2f\n", swf->movieSize.xmin/20.0);
    fprintf(fo, "Height: %.2f\n",(swf->movieSize.ymax-swf->movieSize.ymin)/20.0);
    fprintf(fo, "Y-Offset: %.2f\n", swf->movieSize.ymin/20.0);
    fprintf(fo, "\n");

    while(tag)
    {
	if(swf_isDefiningTag(tag)) {
	    fprintf(fo, "%s%s <%d>\n", prefix, swf_TagGetName(tag), swf_GetDefineID(tag));
	    swf_SetTagPos(tag, 2);
	    dumpTag(fo, prefix, tag);
	} else if(swf_isPseudoDefiningTag(tag)) {
	    fprintf(fo, "%s%s <%d>\n", prefix, swf_TagGetName(tag), swf_GetDefineID(tag));
	    swf_SetTagPos(tag, 2);
	    dumpTag(fo, prefix, tag);
	} else if(tag->id == ST_PLACEOBJECT || tag->id == ST_PLACEOBJECT2) { 
	    SWFPLACEOBJECT po;
	    swf_GetPlaceObject(tag, &po);
	    fprintf(fo, "%s%s <%d>\n",  prefix,swf_TagGetName(tag), po.id);
	    /* for now */
	    swf_SetTagPos(tag, 0);
	    dumpTag(fo, prefix, tag);
	    swf_PlaceObjectFree(&po);
	} else {
	    fprintf(fo, "%s%s\n", prefix, swf_TagGetName(tag));
	    dumpTag(fo, prefix, tag);
	}

	if(tag->id == ST_DEFINESPRITE) {
	    indent+=4;
	    if(indent>32)
		indent = 32;
	    prefix = &whitespace[32-indent];
	} else if(tag->id == ST_END) {
	    indent-=4;
	    if(indent<0)
		indent = 0;
	    prefix = &whitespace[32-indent];
	}

	tag = tag->next;
    }
}

static void readline(FILE*fi, char*line, int maxlen) {
    int pos = 0;
    while(!feof(fi)) {
        if(!fread(&line[pos],1,1,fi))
            break;

	/* cut of preceding whitespace */
	if(pos == 0 && (line[0] == 32 || line[0] == '\t'))
	    continue;

        if(line[pos] == 13 || line[pos]==10)
            break;
	if(pos<maxlen-1)
	    pos++;
    }
    line[pos]=0;

    /* cut off comments */
    char*x=strchr(line,'#');
    if(x) {
        *x=0;
	pos = x-line;
    }

    /* cut off trailing whitespace */
    while(pos>=1 && (line[pos-1]==32 || line[pos-1]==9)) {
	pos--;
	line[pos]=0;
    }
}

int getFloat(char*s)
{
    float x;
    int n;
    while(*s==32 || *s=='\t') s++;
    sscanf(s, "%f%n", &x, &n);
    if(n==0) 
	fprintf(stderr, "Not a float: %s\n", s);
    return x;
}
int getInt(char*s)
{
    int i;
    int n;
    while(*s==32 || *s=='\t') s++;
    sscanf(s, "%d%n", &i, &n);
    if(n==0) 
	fprintf(stderr, "Not an integer: %s\n", s);
    return i;
}
int getTwip(char*x)
{
    return (int)(getFloat(x)*20);
}


static char**lookup;
int swf_TagNameToID(char*name)
{
    int t;
    TAG tag;
    memset(&tag, 0, sizeof(tag));
    if(!lookup) {
	lookup = (char**)malloc(sizeof(char*)*65536);
	for(t=0;t<65536;t++) {
	    tag.id = t;
	    lookup[t] = swf_TagGetName(&tag);
	}
    }
    for(t=0;t<65536;t++) {
	if(lookup[t] && !strcasecmp(name, lookup[t]))
	    return t;
    }
    fprintf(stderr, "Not a tag name: \"%s\"\n", name);
    return -1;
}

void parseFile(FILE*fi, SWF*swf)
{
    TAG*tag = 0;
    char line[1024];
    memset(swf, 0, sizeof(SWF));
    while(1) {
	char*colon = 0;
	readline(fi, line, 1024);
	if(!line[0])
	    break;
	colon = strchr(line, ':');
	if(colon) {
	    int num = colon - line;
	    int n;
	    if(num == 9 && !strncmp(line, "FrameRate", num)) {
		swf->frameRate = getFloat(colon+1)*256;
	    } else if(num == 10 && !strncmp(line, "FrameCount", num)) {
		swf->frameCount = getInt(colon+1);
	    } else if(num == 7 && !strncmp(line, "Version", num)) {
		swf->fileVersion = getInt(colon+1);
	    } else if(num == 5 && !strncmp(line, "Width", num)) {
		int width = getTwip(colon+1);
		swf->movieSize.xmax += width;
	    } else if(num == 6 && !strncmp(line, "Height", num)) {
		int height = getTwip(colon+1);
		swf->movieSize.ymax += height;
	    } else if(num == 8 && !strncmp(line, "X-Offset", num)) {
		int xoffset = getTwip(colon+1);
		swf->movieSize.xmin += xoffset;
		swf->movieSize.xmax += xoffset;
	    } else if(num == 8 && !strncmp(line, "Y-Offset", num)) {
		int yoffset = getTwip(colon+1);
		swf->movieSize.ymin += yoffset;
		swf->movieSize.ymax += yoffset;
	    } else {
		fprintf(stderr, "Ignored line \"%s\"\n", line);
	    }
	}
    }
    while(!feof(fi)) {
	char*s, *tagname;
	char*br;
	readline(fi, line, 1024);
	if(!line[0]) 
	    continue;
	s = strchr(line, ' ');
	br = strchr(line, '|');
	if(!br) {
	    int id = 0;
	    /* DEFINESHAPE <id> ... type line */
	    if(!s) {
		tagname = strdup(line);
	    } else {
		tagname = strdup(line);
		tagname[s-line] = 0;
	    }
	    id = swf_TagNameToID(tagname);
	    free(tagname);
	    if(id<0) {
		fprintf(stderr, "Ignored line \"%s\"\n", line);
		continue;
	    }
	    tag = swf_InsertTag(tag, id);
	    if(!swf->firstTag)
		swf->firstTag = tag;
	} else {
	    /* | 00 34 fe c7 ... type line */
	    char*p = line;
	    int num = 0;
	    int digits = 0;

	    if(!tag) {
		fprintf(stderr, "Discarded unassignable data %s\n", line);
		continue;
	    }

	    while(*p) {
		int n = 0;
		if((*p>='a' && *p<='f') ||
		   (*p>='A' && *p<='F')) {
		    n = 9 + (*p & 15);
		    num = (num<<4) | n;
		    digits++;
		}
		else if(*p>='0' && *p<='9') {
		    n = *p & 15;
		    num = (num<<4) | n;
		    digits++;
		} else if(digits) {
		    swf_SetU8(tag, num);
		    num = 0;
		    digits = 0;
		}
		p++;
	    }
	    if(digits)
		swf_SetU8(tag, num);
	}
    }
}
    
char swf_IsSWF(char*filename)
{
    int fi = open(filename,O_RDONLY|O_BINARY);
    U8 buf[3] = {0,0,0};
    if(fi<0)
	return 0;
    read(fi, buf, 3);
    close(fi);
    if((buf[0] == 'F' || buf[0] == 'C') && buf[1] == 'W' && buf[2] == 'S')
	return 1;
    return 0;
}

int main (int argc,char ** argv)
{ 
    int fi;
    TAG*tag = 0;
    SWF swf;

    processargs(argc, argv);
    
    if(!filename) {
        fprintf(stderr, "You must supply a filename.\n");
        return -1;
    }
    if(swf_IsSWF(filename)) {
	fi = open(filename,O_RDONLY|O_BINARY);
	if (fi<0) { 
	    perror("Couldn't open file: ");
	    exit(1);
	}
	if FAILED(swf_ReadSWF(fi,&swf)) { 
	    fprintf(stderr, "%s is not a valid SWF file or contains errors.\n",filename);
	    close(fi);
	    exit(1);
	}
	dumpFile(&swf, stdout);
	swf_FreeTags(&swf);
    } else {
	SWF newswf;
	FILE*fi = fopen(filename, "rb");
	parseFile(fi, &newswf);
	fclose(fi);
	int f;
	char*sname = output_filename;
	f = open(sname,O_WRONLY|O_CREAT|O_TRUNC|O_BINARY, 0644);
	if FAILED(swf_WriteSWF(f,&newswf)) { 
	    fprintf(stderr, "Unable to write output file: %s\n", sname);
	}
	close(f);
    }

    return 0;
}
